src/virtualenv/interpreters/discovery/py_info.py

```diff
@@ -0,0 +1,256 @@
+            # otherwise fallback to discovery mechanism
+            return self.find_exe_based_of(inside_folder=env_prefix)
+        else:
+            # need original executable here, as if we need to copy we want to copy the interpreter itself, not the
+            # setup script things may be wrapped up in
+            return self.original_executable
+
+    def find_exe_based_of(self, inside_folder):
+        # we don't know explicitly here, do some guess work - our executable name should tell
+        possible_names = self._find_possible_exe_names()
+        possible_folders = self._find_possible_folders(inside_folder)
+        for folder in possible_folders:
+            for name in possible_names:
+                candidate = os.path.join(folder, name)
+                if os.path.exists(candidate):
+                    info = PythonInfo.from_exe(candidate)
+                    keys = {"implementation", "architecture", "version_info"}
+                    if all(getattr(info, k) == getattr(self, k) for k in keys):
+                        return candidate
+        what = "|".join(possible_names)  # pragma: no cover
+        raise RuntimeError("failed to detect {} in {}".format(what, "|".join(possible_folders)))  # pragma: no cover
+
+    def _find_possible_folders(self, inside_folder):
+        candidate_folder = OrderedDict()
+        base = os.path.dirname(self.executable)
+        # following path pattern of the current
+        if base.startswith(self.prefix):
+            relative = base[len(self.prefix) :]
+            candidate_folder["{}{}".format(inside_folder, relative)] = None
+
+        # or at root level
+        candidate_folder[inside_folder] = None
+        return list(candidate_folder.keys())
+
+    def _find_possible_exe_names(self):
+        name_candidate = OrderedDict()
+        for name in [self.implementation, "python"]:
+            for at in range(3, -1, -1):
+                version = ".".join(str(i) for i in self.version_info[:at])
+                for arch in ["-{}".format(self.architecture), ""]:
+                    for ext in EXTENSIONS:
+                        candidate = "{}{}{}{}".format(name, version, arch, ext)
+                        name_candidate[candidate] = None
+        return list(name_candidate.keys())
+
+    _cache_from_exe = {}
+
+    @classmethod
+    def from_exe(cls, exe, raise_on_error=True):
+        key = os.path.realpath(exe)
+        if key in cls._cache_from_exe:
+            result, failure = cls._cache_from_exe[key]
+        else:
+            failure, result = cls._load_for_exe(exe)
+            cls._cache_from_exe[key] = result, failure
+        if failure is not None:
+            if raise_on_error:
+                raise failure
+            else:
+                logging.debug("%s", str(failure))
+        return result
+
+    @classmethod
+    def _load_for_exe(cls, exe):
+        from virtualenv.util.subprocess import subprocess, Popen
```

Comment:
hmmm, this seems to be non-stdlib imports -- is this OK (given the docstring)

Reply:
the from_exe method is an exception as it's only used from within virtualenv 👍